Add new config templates to `tbot` for databases and identity files

timothyb89 opened the pull request:
This adds a number of new config templates building off Teleport's `identityfile` package, adopting several of the formats supported by `tctl auth sign`.

* The `ssh_client` template is now required and will be added automatically in all cases if not specified.
 * A new required `tls_cas` template is added to always export the current Teleport server CAs in a usable format.
 * A new required `identity` template is added to always export an identity file usable with tsh/tctl.
 * New optional `cockroach`, `mongo`, and `tls` templates can export specifically-formatted TLS certs for various databases and apps.

Additionally, this tries to simplify the UX for end-user certificates. First, user-facing `kinds` are removed to reduce confusion
and several config templates are now required. Also, tbot's own formats for CA certs (`tlscacerts`, `sshcacerts`) are no longer written at all to destination dirs since they aren't easily usable by other applications and better alternatives are now available.

The `identity` file can be used in conjunction with `tsh -i` and `tctl -i` to use those tools with a tbot-generated identity. Unfortunately `tsh` only supports identity files for a limited subset of functionality, though SSH use works well. Database, Kubernetes, and app access are currently unsupported through this method.

Additionally some other changes were caught during testing:
 * `botfs` now allows users to specify if files should be opened for reading or for writing; previously, written files were never truncated when opened for writing leading to garbage at the end of files if the length changed. Truncation isn't sane for reading so the two use-cases are now split.
 * The `ssh_client` template was using `os.WriteFile()` directly instead of our filesystem wrapper. It now writes through the wrapper and shouldn't accidentally write files to disk during unit tests.

Fixes #10812

timothyb89 committed:
Add new `identityfile` config template to `tbot`

This adds a new `identityfile` config template to tbot which
generates an identity file from any of the formats supported by
`tctl auth sign`.

It can be used by specifying one or more formats in the configuration
like so:

```yaml
destinations:
  - directory: /foo
    kinds: [ssh, tls]
    configs:
      - identityfile:
          formats: [file]
```

It requires both SSH and TLS certificates to work properly. App,
Kubernetes, and Database certs are unlikely to work as they have
additional cert requirements that will be added in separate PRs.

Multiple formats can be specified, and each will be written to its
own subdirectory within the destination using the name of the format.
The particular files written inside this directory depend on the
particular format selected, but n the above example, this means a
file named `/foo/file/identity` is written.

The files all have an `identity` prefix at the moment. This could be
made configurable if desired.

The `file` format can be used in conjunction with `tsh -i` and
`tctl -i` to use those tools with a tbot-generated identity.

Fixes #10812

timothyb89 commented:
I don't love the UX right now but it does enable a lot of new use-cases in relatively few lines of code. There'll be some growing pains (and breaking changes) if we want to more deeply integrate with `identityfile` and declutter things.

I could see a few alternatives to / iterations on this:
 1. Adopting `identityfile` wholesale, requiring significant (probably breaking) changes to tbot's own filesystem footprint, and lots of refactoring to `identityfile` itself to support our needs
 2. Promote `identityfile` from a config template to it's own destination-level field. It's sort of cleaner, but there's already a fair amount of confusion over `kinds` vs `configs`, so this feels a bit dangerous.
 3. Individually wrap each `identityfile` template into a tbot config template so they become first-class citizens under `configs`. A bit verbose but considerably less work than approach #1. We can also add our own error checking (e.g. requesting a db format cert when you don't actually have a db cert) to further improve UX.

timothyb89 committed:
Make identityfile formats first-class config templates

This promotes most of the important identityfile formats to proper
config templates. User-facing `kinds` are removed to reduce confusion
and several config templates are now required.

 * The `ssh_client` template is now required and will be added
   automatically in all cases if not specified.
 * A new required `tls_cas` template is added to always export
   the current Teleport server CAs in a usable format.
 * A new required `identity` template is added to always export an
   identity file usable with tsh/tctl.
 * New optional `cockroach`, `mongo`, and `tls` templates can export
   specifically-formatted TLS certs for various databases and apps.

Additionally some other changes were caught during testing:
 * `botfs` now allows users to specify if files should be opened for
   reading or for writing; previously, written files were never
   truncated when opened for writing leading to garbage at the end of
   files if the length changed. Truncation isn't sane for reading so
   the two use-cases are now split.

timothyb89 commented:
I've been stalled for a couple of days now on trying to get tsh to accept these new identity files for database access. It's proving to more complex than expected to make this work so I'd like to at least get this PR in to deliver something and will focus on the database multiplexing situation in a follow-up PR.

jakule commented on lib/client/identityfile/identity.go:
```diff
@@ -90,6 +91,40 @@ func (f FormatList) String() string {
+	// permissions if the file is new.
+	WriteFile(name string, data []byte, perm os.FileMode) error
+
+	// Remove removes a file
```
nit:
```suggestion
	// Remove removes a file.
```

jakule commented on tool/tbot/botfs/fs_linux.go:
```diff
@@ -58,12 +58,17 @@ var missingSyscallWarning sync.Once
 // openSecure opens the given path for writing (with O_CREAT, mode 0600)
 // with the RESOLVE_NO_SYMLINKS flag set.
-func openSecure(path string) (*os.File, error) {
+func openSecure(path string, write bool) (*os.File, error) {
+	mode := ReadMode
+	if write {
+		mode = WriteMode
+	}
```
nit: Why not just pass the mode as an argument? You could define a type `type OpenMode int` to make it more explicit.
```suggestion
func openSecure(path string, mode OpenMode) (*os.File, error) {
```
I think that 
```
openStandard(path, WriteMode)
```
is easier to understand.

timothyb89 replied:
Good thought, this is needlessly constrained.

jakule commented on tool/tbot/config/configtemplate_mongo.go:
```diff
@@ -0,0 +1,92 @@
+		return trace.Wrap(err)
+	}
+
+	hostCAs, err := authClient.GetCertAuthorities(ctx, types.HostCA, false)
```
Starting from Teleport 10, Database Access will use Database CA instead of Host CA (if I understand the logic correctly) where Teleport 9.0 and before are still using Host CA.
```suggestion
	hostCAs, err := authClient.GetCertAuthorities(ctx, types.DatabaseCA, false)
```
PR: https://github.com/gravitational/teleport/pull/9593

jakule commented on tool/tbot/config/configtemplate_cockroach.go:
```diff
@@ -0,0 +1,91 @@
+		return trace.Wrap(err)
+	}
+
+	hostCAs, err := authClient.GetCertAuthorities(ctx, types.HostCA, false)
```
The same as mongo. Teleport 10 should use Database CA.

jakule commented on tool/tbot/config/configtemplate_tls.go:
```diff
@@ -0,0 +1,96 @@
+		return trace.Wrap(err)
+	}
+
+	hostCAs, err := authClient.GetCertAuthorities(ctx, types.HostCA, false)
```
Databases should use Database CA here on Teleport 10+.

timothyb89 replied:
Hmmm. This particular template is for general use so I think we'll need to make the CA configurable. The database-specific ones can set that by default, though.

I sort of wonder if we should make specific templates for MySQL / MariaDB / Postgres / etc just to mitigate any confusion.

jakule replied:
This is what I meant. To create a separate implementation for databases as they are using a different CA. Sorry for the confusion.

timothyb89 replied:
No worries! For now I've just made this configurable:

```yaml
configs:
  - tls:
      ca_cert_type: database
```

I suspect we'll be revisiting database-specific configs in the future anyway, so dedicated database-specific configs should be okay as a follow-up PR.

timothyb89 commented on tool/tbot/config/configtemplate.go:
```diff
@@ -94,7 +145,50 @@ func (c *TemplateConfig) CheckAndSetDefaults() error {
+
+	if c.Identity != nil {
+		if err := c.Identity.CheckAndSetDefaults(); err != nil {
+			return trace.Wrap(err)
+		}
+
+		notNilCount++
+	}
+
+	if c.TLS != nil {
+		if err := c.TLS.CheckAndSetDefaults(); err != nil {
+			return trace.Wrap(err)
+		}
+
+		notNilCount++
+	}
+
+	if c.TLSCAs != nil {
+		if err := c.TLSCAs.CheckAndSetDefaults(); err != nil {
+			return trace.Wrap(err)
+		}
+
+		notNilCount++
+	}
+
+	if c.Mongo != nil {
+		if err := c.Mongo.CheckAndSetDefaults(); err != nil {
+			return trace.Wrap(err)
+		}
+
+		notNilCount++
+	}
+
+	if c.Cockroach != nil {
+		if err := c.Cockroach.CheckAndSetDefaults(); err != nil {
+			return trace.Wrap(err)
+		}
+
 		notNilCount++
 	}
```
that... is wildly better, thanks!

timothyb89 replied:
Done, and I also ended up applying this to `GetConfigTemplate()`. It's a much nicer pattern!

timothyb89 committed:
Update lib/client/identityfile/identity.go

Co-authored-by: Jakub Nyckowski <jakub.nyckowski@goteleport.com>

timothyb89 committed:
Address first batch of review comments

Tweaked the `botfs.openStandard` and `botfs.openSecure` functions to
accept a plain file mode, and removed a ton of boilerplate in
`configtemplate.go`.

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/machineid-identityfile

timothyb89 committed:
Merge remote-tracking branch 'origin/timothyb89/machineid-identityfile' into timothyb89/machineid-identityfile

timothyb89 committed:
Fix problematic nil interface check in configtemplate

espadolini commented on tool/tbot/config/configtemplate.go:
```diff
@@ -91,11 +143,31 @@ func (c *TemplateConfig) UnmarshalYAML(node *yaml.Node) error {
+		// Note: this check is fragile and will fail if the templates aren't
+		// all simple pointer types. They are, though, and "correct" solution
+		// is insane, so we'll stick with this.
+		if template == nil || reflect.ValueOf(template).IsNil() {
```
`template == nil` is always false here, right? they're all non-nil `interface { CheckAndSetDefaults() error }`s (which might be nil pointers internally).

timothyb89 replied:
Yep, will remove.

timothyb89 committed:
Clarify comment about `client.Key` DB certs

timothyb89 commented on tool/tbot/config/config_destination.go:
```diff
@@ -63,6 +61,30 @@ func destinationDefaults(dm *DestinationMixin) error {
+// explicitly requested by users. Several configs, including `identity`, `tls`,
+// and `ssh_client`, are always generated (with defaults set, if any) but will
+// not be overridden if already included by the user.
+func (dc *DestinationConfig) addRequiredConfigs() {
```
Assuming you mean "default" vs "required", it's all a bit fuzzy but "default" suggests to me that there are some situations where the templates might not appear. The function makes sure a set of templates is always present even if not/partially/etc specified.

timothyb89 commented on lib/client/identityfile/identity.go:
```diff
@@ -107,11 +142,19 @@ type WriteConfig struct {
 	// overwritten. When false, user will be prompted for confirmation of
 	// overwrite first.
 	OverwriteDestination bool
+	// Writer is the filesystem implementation.
+	Writer ConfigWriter
```
Some templates write multiple files so it seemed a little cleaner this way, I think it'd be awkward to couple the type-specific file prefixing logic with the writer implementation. And I suppose the writer instance is reusable, not that we do that anywhere.

timothyb89 committed:
Address review feedback

 - Use `DatabaseCA` for database specific templates; make the `tls`
   template's CA configurable; write the database CA alongside the
   others.
 - Simplify nil interface check

jakule commented on tool/tbot/config/configtemplate_cockroach.go:
```diff
@@ -0,0 +1,91 @@
+		return trace.Wrap(err)
+	}
+
+	hostCAs, err := authClient.GetCertAuthorities(ctx, types.DatabaseCA, false)
```
nit:
You probably should rename the variable as it holds Database CA now.
```suggestion
	dbCAs, err := authClient.GetCertAuthorities(ctx, types.DatabaseCA, false)
```

timothyb89 replied:
D'oh, I'm getting sloppy with my find-and-replace. Fixed!

jakule commented on tool/tbot/config/configtemplate_mongo.go:
```diff
@@ -0,0 +1,92 @@
+		return trace.Wrap(err)
+	}
+
+	hostCAs, err := authClient.GetCertAuthorities(ctx, types.DatabaseCA, false)
```
The same as above, it's database CA, not host CA.

timothyb89 committed:
Fix outdated var names

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/machineid-identityfile

timothyb89 committed:
Merge branch 'master' into timothyb89/machineid-identityfile

timothyb89 committed:
Make sure to use a random free port for the proxy service in tbot tests

timothyb89 committed:
Resolve lint
